src/v0/destinations/ga4/customMappingsHandler.js

```diff
@@ -0,0 +1,198 @@
+const get = require('get-value');
+const jsonpath = require('rs-jsonpath');
+const {
+  validateEventName,
+  basicValidation,
+  isReservedEventName,
+  addClientDetails,
+  removeReservedParameterPrefixNames,
+  prepareUserConsents,
+  removeInvalidParams,
+  GA4_RESERVED_PARAMETER_EXCLUSION,
+  getGA4CustomParameters,
+  buildDeliverablePayload,
+  GA4_PARAMETERS_EXCLUSION,
+  prepareUserProperties,
+} = require('./utils');
+const { InstrumentationError } = require('@rudderstack/integrations-lib');
+const {
+  removeUndefinedAndNullRecurse,
+  constructPayload,
+  isDefinedAndNotNull,
+  isEmptyObject,
+  removeUndefinedAndNullValues,
+  isHybridModeEnabled,
+  getIntegrationsObj,
+} = require('../../util');
+const { trackCommonConfig, ConfigCategory, mappingConfig } = require('./config');
+
+const findGA4Events = (eventsMapping, event) => {
+  // Find the event using destructuring and early return
+  const validMappings = eventsMapping.filter(
+    (mapping) =>
+      mapping.rsEventName?.trim().toLowerCase() === event.trim().toLowerCase() &&
+      mapping.ga4EventName,
+  );
+  // Return an empty object if event not found
+  return validMappings;
+};
+
+const handleCustomMappings = (message, Config) => {
+  const { eventsMapping } = Config;
+
+  let rsEvent = get(message, 'event');
+  basicValidation(rsEvent);
+
+  const validMappings = findGA4Events(eventsMapping, rsEvent);
+
+  if (validMappings.length === 0) {
+    // Default mapping
+
+    let rawPayload = constructPayload(message, trackCommonConfig);
+
+    const ga4EventPayload = {};
+
+    ga4EventPayload.name = rsEvent;
+
+    // take optional params parameters for custom events
+    ga4EventPayload.params = {
+      ...ga4EventPayload.params,
+      ...constructPayload(message, mappingConfig[ConfigCategory.TrackPageCommonParamsConfig.name]),
+    };
+
+    // all extra parameters passed is incorporated inside params
+    ga4EventPayload.params = getGA4CustomParameters(
+      message,
+      ['properties'],
+      GA4_RESERVED_PARAMETER_EXCLUSION.concat(GA4_PARAMETERS_EXCLUSION),
+      ga4EventPayload,
+    );
+
+    // Prepare GA4 user properties
```

Comment:
can't userProperties preparation & `removeUndefinedAndNullValues` be moved under `boilerplateOperations` as they are agnostic to whether event mappings are defined or not ?